exporter/dynatraceexporter/metrics_exporter.go

```diff
@@ -0,0 +1,181 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dynatraceexporter
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/dynatraceexporter/config"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/dynatraceexporter/serialization"
+)
+
+// NewExporter exports to a Dynatrace MINT API
+func newMetricsExporter(params component.ExporterCreateParams, cfg *config.Config) *exporter {
+	client := &http.Client{}
+
+	return &exporter{params.Logger, cfg, client}
+}
+
+// exporter forwards metrics to a Dynatrace agent
+type exporter struct {
+	logger *zap.Logger
+	cfg    *config.Config
+	client *http.Client
+}
+
+const (
+	maxMetricKeyLen = 250
+)
+
+// Export given CheckpointSet
+func (e *exporter) PushMetricsData(ctx context.Context, md pdata.Metrics) (droppedTimeSeries int, err error) {
+	output := ""
+	dropped := 0
+
+	resourceMetrics := md.ResourceMetrics()
+
+	for i := 0; i < resourceMetrics.Len(); i++ {
+		resourceMetric := resourceMetrics.At(i)
+		if resourceMetric.IsNil() {
+			continue
+		}
+		libraryMetrics := resourceMetric.InstrumentationLibraryMetrics()
+		for j := 0; j < libraryMetrics.Len(); j++ {
+			libraryMetric := libraryMetrics.At(j)
+			if libraryMetric.IsNil() {
+				continue
+			}
+			metrics := libraryMetric.Metrics()
+			for k := 0; k < metrics.Len(); k++ {
```

Comment:
Consider handling processing for loop in a separate method.

Reply:
All that is done in the loop is some nil checking then it delegates to the correct serialization function. I could make the whole for loop body its own function but I think it unnecessary redirection while reading the code.